boa_engine/src/nonmaxu32.rs

```diff
@@ -0,0 +1,180 @@
+//! This module implements `NonMaxU32`
+//! which is known not to equal `u32::MAX`.
+//!
+//! This would be useful for integers like `https://tc39.es/ecma262/#array-index`.
+
+use boa_gc::{unsafe_empty_trace, Finalize, Trace};
+use std::fmt;
+use std::ops::{BitOr, BitOrAssign};
+
+/// An integer that is known not to equal `u32::MAX`.
+///
+/// This enables some memory layout optimization.
+/// For example, `Option<NonMaxU32>` is the same size as `u32`:
+///
+/// ```rust
+/// use std::mem::size_of;
+/// use boa_engine::nonmaxu32::NonMaxU32;
+/// assert_eq!(size_of::<Option<NonMaxU32>>(), size_of::<u32>());
+/// ```
+#[derive(Copy, Clone, Eq, Finalize, PartialEq, Ord, PartialOrd, Hash)]
+#[repr(transparent)]
+pub struct NonMaxU32(u32);
```

Comment:
I would suggest the internal type to be not an enum type and be a structure type:
```rust
pub struct NonMaxU32 {
    inner: u32,
}
```

This would disallow direct editing of the internal structure (having to use getters and setters), and changing the internal type would be API compatible.

Talking about changing the internal type, it might be useful to use a `NonZeroU32`, which is better optimized for space. On the other hand, this would require using a `+1` when storing, and a `-1` when retrieving, which could cause a small performance penalty. We should test both options.

Reply:
Sorry for late reply 🙏 Thank you for your suggestion! I will try to work on a fix.

> On the other hand, this would require using a +1 when storing, and a -1 when retrieving, which could cause a small performance penalty.

This sounds a good idea to me. I experimented a bit using XOR to save with `NonZeroU32` like [the nonmax crate does](https://github.com/LPGhatguy/nonmax/blob/d1d4259d4c86646e4e4fce5078bdd216b7042f65/src/lib.rs#L128-L133) last week but I realized we would need to `rev()` the iterator because the xor'ed value would be in reversed order. So, I wonder just +1 and -1 would be easier.

I will try it more and share more information here or on Discord 🙏  Thank you!

Reply:
I think I'm ok with either approach (-1/+1 or XOR).